File: sources/com/s8/io/bytes/api/ByteOutflow.java
package com.s8.io.bytes.api;

import java.io.IOException;

import com.s8.alpha.objects.S8Address;



/**
 * 
 * <p>
 * /!\ All array or String lengths are encoded as UInt32 to reflect their
 * maximum possible length in JAVA.
 * </p>
 * 
 * @author pc
 *
 */
public interface ByteOutflow {


	/**
	 * Write bytes array. Auto-feed underlying ByteBuffer as necessary.
	 * @param bytes
	 * @throws IOException
	 */
	public void putByteArray(byte[] bytes) throws IOException;

	
	/**
	 * 
	 * @param array
	 * @param offset
	 * @param length
	 * @throws IOException 
	 */
	public void putByteArray(byte[] array, int offset, int length) throws IOException;

	
	/**
	 * 
	 * @param head
	 * @throws IOException
	 */
	/*
	public default void putBytes(QxBytes head) throws IOException{
		while(head!=null) {
			putByteArray(head.bytes, head.offset, head.length);
			head = head.next;
		}
	}
	*/
	

	/**
	 * Auto-feed underlying ByteBuffer as necessary.
	 * @param flags
	 * @throws IOException
	 */
	public void putFlags8(boolean[] flags) throws IOException;


	/**
	 * Hack for js part
	 * 
	 * @param value
	 * @throws IOException
	 */
	public default void putBoolean(boolean value) throws IOException {
		putUInt8(value?((byte) 32):((byte) 33));
	}
	

	/**
	 * Directly put byte
	 * @param b
	 * @throws IOException
	 */
	void putByte(byte b) throws IOException;
	


	/**
	 * Auto-feed underlying ByteBuffer as necessary.
	 * @param value
	 * @throws IOException
	 */
	public void putUInt8(int value) throws IOException;


	/**
	 * Auto-feed underlying ByteBuffer as necessary.
	 * @param value
	 * @throws IOException
	 */
	public void putUInt16(int value) throws IOException;
	


	/**
	 * Auto-feed underlying ByteBuffer as necessary.
	 * @param value
	 * @throws IOException
	 */
	public void putInt16(short value) throws IOException;



	/**
	 * Auto-feed underlying ByteBuffer as necessary.
	 * @param value
	 * @throws IOException
	 */
	public void putUInt32(int value) throws IOException;
	
	/**
	 * Auto-feed underlying ByteBuffer as necessary.
	 * This format is adequate to send JAVA long to JS (since JS MAX_INTEGER is 2^53).
	 * @param value
	 * @throws IOException
	 */
	public void putUInt53(long value) throws IOException;

	
	/**
	 * put Uint32 using extensible mechanism to save bytes.
	 * Mechanism is compliant with natural JAVA and network Big-Endianness.
	 * @param value
	 * @throws IOException
	 */
	public void putUInt(int value) throws IOException;
	
	
	/**
	 * Auto-feed underlying ByteBuffer as necessary.
	 * @param value
	 * @throws IOException
	 */
	public void putInt32(int value) throws IOException;


	/**
	 * Auto-feed underlying ByteBuffer as necessary.
	 * @param value
	 * @throws IOException
	 */
	public void putInt64(long value) throws IOException;



	/**
	 * Auto-feed underlying ByteBuffer as necessary.
	 * @param value
	 * @throws IOException
	 */
	public void putFloat32(float value) throws IOException;




	/**
	 * Auto-feed underlying ByteBuffer as necessary.
	 * @param value
	 * @throws IOException
	 */
	public void putFloat64(double value) throws IOException;



	/**
	 * String is encoded in UTF8.
	 * 
	 * @return String
	 * @throws IOException 
	 */
	public void putL32StringUTF8(String str) throws IOException;

	
	/**
	 * 
	 * @param str
	 * @throws IOException
	 */
	public void putL8StringASCII(String str) throws IOException;

	
	
	/**
	 * 
	 * @param strs
	 * @throws IOException
	 */
	public default void putS8Address(S8Address address) throws IOException {
		putByteArray(address.bytes);
	}
	
	/**
	 * 
	 * @param strs
	 * @throws IOException
	 */
	public void putS8Index(long key) throws IOException;
}

File: web-helium/ByteOutflow.js


/**
 * <p>
 * Unified in-flow of bytes with getters for standard primitives.
 * </p>
 * 
 * <p>
 * /!\ All array or String lengths are encoded as UInt32 to reflect their
 * maximum possible length in JAVA.
 * </p>
 * 
 * @author pc
 *
 */
export class ByteOutflow {

	/**
	 * 
	 * @param {ArrayBuffer} arraybuffer 
	 */
	constructor(arraybuffer){
		this.arraybuffer = arraybuffer;
		this.view = new DataView(arraybuffer);
		this.textEncoder_ASCII = new TextEncoder('ascii');
		this.textEncoder_UTF8 = new TextEncoder('utf8');
		this.offset = 0;
	}

	
	putByte(value){
		this.view.setUint8(this.offset, value);
		this.offset+=1;
	}


	putUInt8(value){
		this.view.setUint8(this.offset, value);
		this.offset+=1;
	}

	putUInt16(value){
		this.view.setUint16(this.offset, value);
		this.offset+=2;
	}

	getInt16(value){
		this.view.setInt16(this.offset, value);
		this.offset+=2;
	}

	putUInt32(value){
		this.view.setUInt32(this.offset, value);
		this.offset+=4;
	}

	putInt32(value){
		this.view.setInt32(this.offset, value);
		this.offset+=4;
	}

	putUInt64(value){
		this.view.setUInt64(this.offset, value);
		this.offset+=8;
	}
	
	putInt64(value){
		this.view.setInt64(this.offset, value);
		this.offset+=8;
	}

	putFloat32(value){
		this.view.setFloat32(this.offset, value);
		this.offset+=4;
	}

	putFloat64(value){
		this.view.setFloat64(this.offset, value);
		this.offset+=8;
	}

	putL8StringASCII(value){
		let stringBuffer = this.textEncoder_ASCII.encode(value);
		let length = stringBuffer.length;
		this.view.setUint8(this.offset, length);
		this.offset+=1;
		new Uint8Array(this.arraybuffer).set(stringBuffer, this.offset);
		this.offset+=length;
	}

	putL32StringUTF8(value){
		let stringBuffer = this.textEncoder_UTF8.encode(value);
		let length = stringBuffer.length;
		this.view.setUint32(this.offset, length);
		this.offset+=4;
		new Uint8Array(this.arraybuffer).set(stringBuffer, this.offset);
		this.offset+=length;
	}

}